### ISSUE TITLE
Add command line option to stack script to open it in ghci instead of running it

### ISSUE THREAD

**Issue by lehmacdj** [commercialhaskell/stack#5354]
Make sure that you are using the latest release (currently stack-2.3.1).
See the [upgrade instructions](http://docs.haskellstack.org/en/stable/install_and_upgrade/#upgrade) to upgrade.

Please state as clearly as possible what feature you are recommending,
with motivation for the reason such a feature should be added.

**Comment by friedbrice** [comment]
As a proof-of-concept, I recently created a cross-platform (Windows, Linux, and Mac) tool that does something similar. You specify a resolver and a your package dependencies in a comment at the top of your file, and the tool invokes Stack to either run the file as a script, load the file in GHCi (or Ghcid), or compile the file at the user's option. The intent is to make it easy for casual programmers to use Haskell for one-off tasks without needing to create a full Stack project. You can find it here: http://hackage.haskell.org/package/runhs

I wonder if it'd be worthwhile to add such a feature directly to Stack. I would rather see it there instead of as a separate tool. If I were to make a PR to add this feature to Stack, would the Stack maintainers approve?

**Comment by lehmacdj** [comment]
This looks cool, but I think it would make more sense to integrate with the existing `stack script` functionality. That way we don't need to have two different formats for headers for files.

**Comment by friedbrice** [comment]
That's certainly the case. The problem with Stack script is that they aren't cross-platform. Case in point:

```haskell
#!/usr/bin/env stack
-- stack script --resolver nightly --package text
{-# LANGUAGE OverloadedStrings #-}
import Data.Text.IO as T
main = T.putStrLn "Hey."
```

There's no way to run this script in Windows without changing the source code (similar to how "stack script" is hardcoded in the source code under Stacks current script implementation).

The goal here is to provide a single format for front matter that is cross-platform and cross-interpreter (repl, script, or compiler), so that the same source code could be used in any combination of those contexts. This subsumes Stacks current script implementation, so if it would suite your needs (and if the Stack devs agree) it might even make sense to add this feature and deprecate the current scripts implementation.

If that doesn't sound like it'd suite your needs, I'll create a separate issue.

**Comment by lehmacdj** [comment]
> There's no way to run this script in Windows without changing the source code.

I suppose by this you are referring to the shebang. I'm pretty sure that even on windows (don't have Windows easily accessible to test though) you can run this using `stack script Hey.hs` if this is in a file named "Hey.hs".

In my opinion replacing the existing functionality with `runhs` would be a minor regression because it would forbid using it as a script interpreter on Linux, and running such a script would always require invoking runhs. 

Unless of course `runhs` also supports being called from a shebang, in which case I suppose it would be advantageous, but the loss of backwards compatibility with existing stack scripts seems like a minor nuissance as well.

**Comment by friedbrice** [comment]
> I'm pretty sure that even on windows (don't have Windows easily accessible to test though) you can run this using stack script Hey.hs if this is in a file named "Hey.hs".

I guess I was premature in saying "no way." What I meant is that it can't be made executable and then invoked at the command line, but as you pointed out, there's a way around that by invoking `stack`.

Since the primary goal of Stack is reproducible builds, I agree with you that we would not want to see the Stack devs remove the ability to use Stack as a script interpreter. Along those same lines, we wouldn't want them to stop parsing the existing front matter format. I think you're idea, of adding a `stack script --ghci MyFile.hs` is probably the way to go.

**Comment by Martinsos** [comment]
This would be great, I am trying to use Dante with stack scripts and it seems that this is what is missing to get it working: https://github.com/jyp/dante/issues/165https://github.com/jyp/dante/issues/165 .

@lehmacdj do you still plan to make a PR for this? I would love to help or maybe even do it on my own - I am not sure how much work will this be.

**Comment by lehmacdj** [comment]
I would probably still be willing to work on this, but I would prefer to have blessing from a contributor to stack before getting started, and agreement on what the design should be.

I don’t expect this will be a terribly large amount of work.

**Comment by Martinsos** [comment]
@lehmacdj makes sense! My 2 cents: `stack script --ghci` (or `stack script --repl`) sounds completely reasonable to me.
Since it has been ~9 months since you opened this issue and no contributors have responded, I am not sure if it is reasonable to expect them to get to it soon - probably they are overwhelmed with other issues/work. Not sure what to do thought - is it polite to ping them directly? If we can't get their attention I would consider doing a PR nevertheless and hoping that PR will get more attention than just an issue. If it is not much work it should not be a huge waste in case PR never goes through.

**Comment by Martinsos** [comment]
I was just looking at support for stack scripts in haskell-language-server, and it seems they can't implement support for it until Stack adds support for running scripts in ghci (which is this issue about).

This is mentioned here: https://github.com/mpickering/hie-bios/issues/217 .

Also interesting: in the issue above, they say that before `stack repl script.hs` was running the script in ghci, but now it does not any more (instead it ignores the "script" stuff and runs it as a normal module).
Maybe this was intentional, because there was no way then to run the file just as a module via `stack repl` (although I am not sure why would that be desirable) and it was unexpected behaviour?

Anyway, figuring this out would enable both dante and hls to work with stack scripts.

What I am not sure about is if it should be done as `stack repl script.hs` or `stack script --ghci` -> it seems that `stack repl script.hs` would work with hls and dante immediatelly, while `stack script --ghci` would need some adjusting?

**Comment by Martinsos** [comment]
This issue might be useful for implementing this: https://github.com/commercialhaskell/stack/issues/5525 .

**Comment by tannerlegvold** [comment]
For what its worth, I (a person somewhat new to Haskell) recently [asked](https://www.reddit.com/r/haskell/comments/r6dnmn/monthly_hask_anything_december_2021/hps4asq/) how to do this on the Haskell subreddit. Eventually I accepted that Stack can't do this (right now) but discovered `runhs` and am using that now. So there's demand for this at least amongst beginners.
